File: app/controllers/integration/ReportAPIController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.integration;

import be.objectify.deadbolt.java.actions.SubjectNotPresent;
import controllers.base.BaseController;
import io.ebean.DB;
import io.ebean.Query;
import io.ebean.text.PathProperties;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import models.base.GeneratedIdentityModel;
import models.enrolment.ExamEnrolment;
import models.exam.Exam;
import models.facility.Software;
import org.joda.time.DateTime;
import org.joda.time.Interval;
import org.joda.time.format.ISODateTimeFormat;
import play.mvc.Result;

public class ReportAPIController extends BaseController {

    @SubjectNotPresent
    public Result getExamEnrolments(Optional<String> start, Optional<String> end) {
        PathProperties pp = PathProperties.parse(
            "(id, enrolledOn, noShow, " +
                "reservation(id, machine(id, name, room(name, roomCode)), startAt, endAt, externalReservation(orgName)), " +
                "examinationEventConfiguration(examinationEvent(start)), " +
                "exam(id, course(name, code, credits, identifier, courseImplementation, " +
                "gradeScale(description, displayName), organisation(code, name)), " +
                "softwares(name), duration, examType(type), creditType(type), executionType(type), " +
                "implementation, trialCount, answerLanguage, periodStart, periodEnd, " +
                "examParticipation(started, ended, id))" +
                ")"
        );

        Query<ExamEnrolment> query = DB.find(ExamEnrolment.class);
        pp.apply(query);

        List<ExamEnrolment> participations = query
            .where()
            .or()
            .ne("exam.state", Exam.State.PUBLISHED)
            .eq("noShow", true)
            .endOr()
            .or()
            .isNotNull("reservation.machine")
            .isNotNull("reservation.externalReservation")
            .isNotNull("examinationEventConfiguration")
            .endOr()
            .findList()
            .stream()
            .filter(ee -> filterByDate(ee, start, end))
            .toList();

        /* Relations from exam to software exist on parent exam, therefore fetch parent exam IDs separately */
        Set<Long> parentExamIds = participations
            .stream()
            .filter(participation -> participation.getExam() != null && participation.getExam().getParent() != null)
            .map(participation -> participation.getExam().getParent().getId())
            .collect(Collectors.toSet());

        Map<Long, List<Software>> softwareByExam = DB.find(Exam.class)
            .fetch("softwares", "name")
            .where()
            .idIn(parentExamIds)
            .findList()
            .stream()
            .collect(Collectors.toMap(GeneratedIdentityModel::getId, Exam::getSoftwareInfo));

        /* Set software lists to child exams */
        participations
            .stream()
            .filter(
                participation ->
                    participation.getExam() != null &&
                    participation.getExam().getParent() != null &&
                    softwareByExam.containsKey(participation.getExam().getParent().getId())
            )
            .forEach(participation -> {
                Long parentId = participation.getExam().getParent().getId();
                List<Software> software = softwareByExam.get(parentId);
                if (software != null && !software.isEmpty()) {
                    participation.getExam().setSoftwareInfo(software);
                }
            });

        return ok(participations, pp);
    }

    private boolean filterByDate(ExamEnrolment enrolment, Optional<String> start, Optional<String> end) {
        var min = start.orElse(new DateTime(0L).toDateTimeISO().toString());
        var max = end.orElse(new DateTime(Long.MAX_VALUE).toDateTimeISO().toString());
        var startDate = ISODateTimeFormat.dateTimeParser().parseDateTime(min);
        var endDate = ISODateTimeFormat.dateTimeParser().parseDateTime(max);
        var range = new Interval(startDate, endDate);
        if (enrolment.getReservation() != null) {
            var reservation = enrolment.getReservation();
            var period = new Interval(reservation.getStartAt(), reservation.getEndAt());
            return range.contains(period);
        } else if (enrolment.getExaminationEventConfiguration() != null) {
            var event = enrolment.getExaminationEventConfiguration().getExaminationEvent();
            var duration = enrolment.getExam().getDuration();
            var period = new Interval(event.getStart(), event.getStart().plusMinutes(duration));
            return range.contains(period);
        }
        return true; // if we have no date info on this object, just show it
    }
}

File: app/controllers/integration/ReportAPIController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.integration

import io.ebean.DB
import io.ebean.text.PathProperties
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.enrolment.ExamEnrolment
import models.exam.Exam
import org.joda.time.format.ISODateTimeFormat
import org.joda.time.{DateTime, Interval}
import play.api.mvc.{Action, AnyContent, BaseController, ControllerComponents}
import security.scala.Auth.subjectNotPresent

import javax.inject.Inject
import scala.concurrent.ExecutionContext
import scala.jdk.CollectionConverters.*

class ReportAPIController @Inject() (
    val controllerComponents: ControllerComponents,
    implicit val ec: ExecutionContext
) extends BaseController
    with DbApiHelper
    with JavaApiHelper:

  def getExamEnrolments(start: Option[String], end: Option[String]): Action[AnyContent] =
    Action.andThen(subjectNotPresent) { _ =>
      val pp = PathProperties.parse(
        "(id, enrolledOn, noShow, " +
          "reservation(id, machine(id, name, room(name, roomCode)), startAt, endAt, externalReservation(orgName)), " +
          "examinationEventConfiguration(examinationEvent(start)), " +
          "exam(id, course(name, code, credits, identifier, courseImplementation, " +
          "gradeScale(description, displayName), organisation(code, name)), " +
          "softwares(name), duration, examType(type), creditType(type), executionType(type), " +
          "implementation, trialCount, answerLanguage, periodStart, periodEnd, " +
          "examParticipation(started, ended, id))" +
          ")"
      )

      val query = DB.find(classOf[ExamEnrolment])
      pp.apply(query)

      val participations = query
        .where()
        .ne("exam.state", Exam.State.PUBLISHED)
        .or()
        .isNotNull("reservation.machine")
        .isNotNull("reservation.externalReservation")
        .isNotNull("examinationEventConfiguration")
        .endOr()
        .list
        .filter(ee => filterByDate(ee, start, end))

      // Relations from exam to software exist on the parent exam. Therefore, fetch parent exam IDs separately
      val parentExamIds = participations.flatMap { participation =>
        Option(participation.getExam)
          .flatMap(exam => Option(exam.getParent))
          .map(_.getId)
      }.toSet

      val softwareByExam = DB
        .find(classOf[Exam])
        .fetch("softwares", "name")
        .where()
        .idIn(parentExamIds.map(id => Long.box(id)).asJava)
        .list
        .map(exam => exam.getId -> exam.getSoftwareInfo.asScala.toList)
        .toMap

      // Set software lists to child exams
      participations.foreach { participation =>
        for
          exam     <- Option(participation.getExam)
          parent   <- Option(exam.getParent)
          software <- softwareByExam.get(parent.getId) if software.nonEmpty
        do exam.setSoftwareInfo(software.asJava)
      }

      Ok(participations.asJson(pp))
    }

  private def filterByDate(enrolment: ExamEnrolment, start: Option[String], end: Option[String]): Boolean =
    val min       = start.getOrElse(new DateTime(0L).toDateTimeISO.toString)
    val max       = end.getOrElse(new DateTime(Long.MaxValue).toDateTimeISO.toString)
    val startDate = ISODateTimeFormat.dateTimeParser().parseDateTime(min)
    val endDate   = ISODateTimeFormat.dateTimeParser().parseDateTime(max)
    val range     = new Interval(startDate, endDate)

    Option(enrolment.getReservation)
      .map { reservation =>
        val period = new Interval(reservation.getStartAt, reservation.getEndAt)
        range.contains(period)
      }
      .orElse {
        Option(enrolment.getExaminationEventConfiguration).map { config =>
          val event    = config.getExaminationEvent
          val duration = enrolment.getExam.getDuration
          val period   = new Interval(event.getStart, event.getStart.plusMinutes(duration))
          range.contains(period)
        }
      }
      .getOrElse(true) // if we have no date info on this object, just show it
